quota/etcd/storage/quota_storage.go

```diff
@@ -0,0 +1,387 @@
+	if err != nil {
+		glog.Fatalf("bad global pattern: %v", err)
+	}
+	treesPattern, err = regexp.Compile("quotas/trees/\\d+/(read|write)/config")
+	if err != nil {
+		glog.Fatalf("bad trees pattern: %v", err)
+	}
+	usersPattern, err = regexp.Compile("quotas/users/[^/]+/(read|write)/config")
+	if err != nil {
+		glog.Fatalf("bad users pattern: %v", err)
+	}
+}
+
+// QuotaStorage is the interface between the etcd-based quota implementations (quota.Manager and
+// RPCs) and etcd itself.
+type QuotaStorage struct {
+	Client *clientv3.Client
+}
+
+// UpdateConfigs creates or updates the supplied configs in etcd.
+// If no config exists, the current config is assumed to be an empty storagepb.Configs proto.
+// The update function allows for mask-based updates and ensures a single-transaction
+// read-modify-write operation.
+// If reset is true, all specified configs will be set to their max number of tokens. If false,
+// existing quotas won't be modified, unless the max number of tokens is downgraded, in which case
+// the new ceiling is enforced.
+// Newly created quotas are always set to max tokens, regardless of the reset parameter.
+func (qs *QuotaStorage) UpdateConfigs(ctx context.Context, reset bool, update func(*storagepb.Configs)) (*storagepb.Configs, error) {
+	if update == nil {
+		return nil, errors.New("update function required")
+	}
+
+	var updated *storagepb.Configs
+	_, err := concurrency.NewSTMSerializable(ctx, qs.Client, func(s concurrency.STM) error {
+		previous, err := getConfigs(s)
+		if err != nil {
+			return err
+		}
+		// Take a deep copy of "previous". It's pointers all the way down, so it's easier to just
+		// unmarshal it again. STM has the key we just read and it should be exactly the same as
+		// previous, so it's ok to ignore the error...
+		updated, _ = getConfigs(s)
+
+		// ... but let's sanity check that the configs match, just in case.
+		if !proto.Equal(previous, updated) {
+			return errors.New("verification failed: previous != updated")
+		}
+
+		update(updated)
+		if err := validate(updated); err != nil {
+			return err
+		}
+
+		if !proto.Equal(previous, updated) {
+			pb, err := proto.Marshal(updated)
+			if err != nil {
+				return err
+			}
+			s.Put(configsKey, string(pb))
+		}
+
+		now := timeNow()
+		for _, cfg := range updated.Configs {
+			// Make no distinction between enabled and disabled configs here. Get/Peek/Put are
+			// prepared to handle it, and recording the bucket as if it were enabled allows us to
+			// take advantage of the already-existing reset and downgrade logic.
+			key := bucketKey(cfg)
+
+			var prev *storagepb.Config
+			for _, p := range previous.Configs {
+				if cfg.Name == p.Name {
+					prev = p
+					break
+				}
+			}
+
+			switch {
+			case prev == nil || prev.State == storagepb.Config_DISABLED || reset: // new bucket
+				bucket := &storagepb.Bucket{
+					Tokens: cfg.MaxTokens,
+					LastReplenishMillisSinceEpoch: now.UnixNano() / 1000000,
```

Comment:
Do you mean `now.UnixNano() / int64(time.Millisecond)`? I find that more confusing than just using a plain number. Nanos/Millis happens to work because of the way Go represents numbers, but Micros/Millis wouldn't, for example. What we want here is just a number.

I did change the numbers to the `1eX` format.